website/src/app/blog/how-dns-works-in-firezone/readme.mdx

```diff
@@ -1,66 +1,373 @@
+generates a matching sentinel address in the `100.100.111.0/24` and
+`fd00:2021:1111:8000:100:100:111:0/120` ranges for IPv4 and IPv6 resolvers,
+respectively. This is why you'll often see `nameserver 100.100.111.1` as one of
+your upstream resolvers in `/etc/resolv.conf` while the Client is connected.
+
+A nice side effect of this one-to-one mapping approach is that it won't affect
+the selection algorithm your operating system uses to pick healthy nameservers
+-- if one is down, the corresponding sentinel address will be unresponsive, and
+the operating system will pick another, responsive sentinel to use instead.
+
+As we were building it, we thought of another useful feature: Instead of simply
+using the system's default resolvers to forward queries to, we can use ones set
+by an admin in the admin portal instead. This can be used, for example, to
+configure DNS filtering provider to
+[block malicious DNS queries](/kb/use-cases/secure-dns) across your workforce.
+Or you could point it to your organization's internal nameservers to resolve
+internal services in a more traditional Split DNS configuration.
+
+#### Generating the mappings
+
+Ok, so that covers how queries are intercepted, but how does the stub resolver
+actually generate the dummy IP addresses for your Firezone Resources? Let's walk
+through an example to illustrate.
+
+<p className="text-sm text-center mx-auto italic">
+  Figure 9: Generating dummy IP addresse (TODO)
+</p>
+
+1. An admin defines a Resource (say `*.slack.com`) and then adds a corresponding
+   Policy to grant access.
+1. All connected Clients allowed by the Policy will immediately receive the new
+   Resource definition.
+1. The Client then configures the stub resolver to begin intercepting queries
+   for `*.slack.com` in order to forward to the Policy Engine.
+1. When it sees a matching one, the stub resolver forwards the query to the
+   Policy Engine, which then reauthorizes the query and finds a healthy Gateway
+   to resolve it.
+1. The Gateway resolves the query, remembering the Client that just asked it,
+   and then returns all of the resolved IP addresses to the stub resolver in the
+   Client.
+1. The stub resolver then generates a unique IP address for each resolved IP
+   address, adds them to our host's routing table, and returns these to the
+   application that made the query.
+1. The application then begins sending packets to the dummy IP address, where
+   they're routed through a secure WireGuard tunnel to Gateway we just resolved
+   the query with.
+
+Similar to the way sentinel generation works above, the stub resolver generates
+a single IPv4 or IPv6 address for _each_ resolved IP address the Gateway finds,
+picking a sequential address from the `100.96.0.0/11` and
+`fd00:2021:1111:8000::/107` ranges to use for the mapping respectively. This
+ensures that things like `NXDOMAIN` behavior and round-robin DNS continue to
```

Comment:
```suggestion
ensures that things like timeout behavior and round-robin DNS continue to
```

most stub resolvers stop trying when `NXDOMAIN` is returned since they assume the same domains are present in all servers